state/protocol/protocol_state/epochs/statemachine.go

```diff
@@ -0,0 +1,300 @@
+	// As a result of this operation protocol state for next epoch will be committed.
+	// Returned boolean indicates if event triggered a transition in the state machine or not.
+	// Implementors must never return (true, error).
+	// Expected errors indicating that we are leaving the happy-path of the epoch transitions
+	//   - `protocol.InvalidServiceEventError` - if the service event is invalid or is not a valid state transition for the current protocol state.
+	//     CAUTION: the HappyPathStateMachine is left with a potentially dysfunctional state when this error occurs. Do NOT call the Build method
+	//     after such error and discard the HappyPathStateMachine!
+	ProcessEpochCommit(epochCommit *flow.EpochCommit) (bool, error)
+
+	// EjectIdentity updates identity table by changing the node's participation status to 'ejected'.
+	// Should pass identity which is already present in the table, otherwise an exception will be raised.
+	// Expected errors during normal operations:
+	// - `protocol.InvalidServiceEventError` if the updated identity is not found in current and adjacent epochs.
+	EjectIdentity(nodeID flow.Identifier) error
+
+	// TransitionToNextEpoch discards current protocol state and transitions to the next epoch.
+	// Epoch transition is only allowed when:
+	// - next epoch has been set up,
+	// - next epoch has been committed,
+	// - candidate block is in the next epoch.
+	// No errors are expected during normal operations.
+	TransitionToNextEpoch() error
+
+	// View returns the view that is associated with this StateMachine.
+	// The view of the StateMachine equals the view of the block carrying the respective updates.
+	View() uint64
+
+	// ParentState returns parent protocol state that is associated with this StateMachine.
+	ParentState() *flow.RichProtocolStateEntry
+}
+
+// StateMachineFactoryMethod is a factory method to create state machines for evolving the protocol's epoch state.
+// Currently, we have `HappyPathStateMachine` and `FallbackStateMachine` as StateMachine
+// implementations, whose constructors both have the same signature as StateMachineFactoryMethod.
+type StateMachineFactoryMethod func(candidateView uint64, parentState *flow.RichProtocolStateEntry) (StateMachine, error)
+
+// EpochStateMachine is a hierarchical state machine that encapsulates the logic for protocol-compliant evolution of Epoch-related sub-state.
+// EpochStateMachine processes a subset of service events that are relevant for the Epoch state, and ignores all other events.
+// EpochStateMachine delegates the processing of service events to an embedded StateMachine,
+// which is either a HappyPathStateMachine or a FallbackStateMachine depending on the operation mode of the protocol.
+// It relies on Key-Value Store to commit read the parent state and to commit the updates to the Dynamic Protocol State.
+type EpochStateMachine struct {
+	candidate                        *flow.Header
```

Comment:
⚠️ causal circle. Header of the output block is not available while the primary builds the block initially. 

Specifically, this line https://github.com/onflow/flow-go/blob/d69920a457a1dbde4c2a01485414d8645a1b810f/state/protocol/protocol_state/epochs/statemachine.go#L170 is not generally correct in the context we execute this logic